File: Dictionary.rb
require 'set'
require_relative 'Word'

class Dictionary
  attr_reader :all_words, :word_length, :letter_positions, :contains_letter
  def initialize(filename)

    @all_words = {}
    # index of all words in dictionary
    # { "CAT" => Word(CAT) }

    @word_length = {}
    # all words sorted into buckets based on their length
    # {
    #   3 => <Set: Word(CAT), Word(BAG)>,
    #   4 => <Set: Word(BEAR)>
    # }

    @letter_positions = []
    # trie of words with letter at positions
    # [
    #   {
    #     B => <Set: Word(BAT), Word(BAG)>,
    #     C => <Set: Word(CAT)>
    #   },
    #   {
    #     A => <Set: Word(CAT), Word(BAT), Word(BAG)>
    #   },
    #   {
    #     G => <Set: Word(BAG)>,
    #     T => <Set: Word(CAT), Word(BAT)>
    #   }
    # ]
    #

    @contains_letter = {}
    # buckets of words based on letters they contain
    # {
    #   A => <Set: Word(CAT), Word(BAT), Word(BAG)>,
    #   B => <Set: Word(BAT), Word(BAG)>,
    #   G => <Set: Word(BAG)>,
    #   T => <Set: Word(CAT), Word(BAT)>
    # }

    # TODO: error handling for loading file
    words = File.new(filename)

    words.each_line do |word|
      # create a new word object from the word
      word = Word.new(word.strip)

      # add word to all words
      @all_words[word.word] = word

      # add word to length bucket
      @word_length[word.length] ||= Set.new
      @word_length[word.length] << word

      # add letters to position bucket
      (0...(word.length)).each do |i|
        ch = word.word[i]
        @letter_positions[i] ||= {}
        @letter_positions[i][ch] ||= Set.new
        @letter_positions[i][ch] << word
      end

      # add to contains letter
      word.unique_letters.each do |ch|
        @contains_letter[ch] ||= Set.new
        @contains_letter[ch] << word
      end

    end # words.each_line

  end # initialize

  def find(string)
    @all_words[string]
  end

  def getWordsWithLetterAtPosition(letter, position)
    @letter_positions[position][letter] || Set.new
  end

  def getWordsWithLetter(letter)
    @contains_letter[letter] || Set.new
  end

  def getWordWithLenth(length)
    @word_length[length] || Set.new
  end

end

File: Dictionary.java
import java.util.*;
import java.io.*;

public class Dictionary {
    
    public final HashMap<String, Word> allWords = new HashMap<>();
    public final HashMap<Integer, HashSet<Word>> wordLength = new HashMap<>();
    public final ArrayList<HashMap<Character, HashSet<Word>>> letterPositions = new ArrayList<>();
    public final HashMap<Character, HashSet<Word>> containsLetter = new HashMap<>();
    
    public Dictionary(String filepath) {
        
        try {
            FileReader fr = new FileReader(filepath);
            BufferedReader br = new BufferedReader(fr);
            String line = br.readLine();
            
            while (line != null) {
                // create word object from string
                if (line.length() > 0) {
                    Word word = new Word(line);
                
                    // insert word object into allwords
                    allWords.put(word.word, word);

                    // insert word object into length bucket
                    if (wordLength.get(word.length) == null) {
                        wordLength.put(word.length, new HashSet<Word>());
                    }
                    wordLength.get(word.length).add(word);

                    // add every character into char-position bucket
                    // add set to map
                    for (int i = 0; i < word.length; i++) {
                        if (i >= letterPositions.size()) {
                            // create map for positions that do not have maps
                            letterPositions.add(i, new HashMap<Character, HashSet<Word>>());
                        }

                        HashMap<Character, HashSet<Word>> hm = letterPositions.get(i);
                        // create set if it doesn't exist
                        if (hm.get(word.word.charAt(i)) == null) {
                            hm.put(word.word.charAt(i), new HashSet<Word>());
                        }
                        // add word to set
                        hm.get(word.word.charAt(i)).add(word);
                    }

                    Iterator<Character> iter = word.uniqueLetters.iterator();
                    while (iter.hasNext()) {
                        Character ch = iter.next();
                        if (containsLetter.get(ch) == null) {
                            containsLetter.put(ch, new HashSet<Word>());
                        }
                        containsLetter.get(ch).add(word);
                    }
                }
                
                // read next line
                line = br.readLine();
            } 
        } catch(IOException e) {
            e.printStackTrace();
            System.exit(1);
        }
    } // public Dictionary
    
    public Word find(String word) {
        return allWords.get(word);
    }
    
    public Set<Word> getWordsWithLetterAtPosition(Character ch, int position) {
        Set<Word> set = letterPositions.get(position).get(ch);
        if (set == null) {
            return new HashSet<>();
        } else {
            return set;
        }
    }
    
    public Set<Word> getWordsWithLetter(Character ch) {
        Set<Word> set = containsLetter.get(ch);
        if (set == null) {
            return new HashSet<>();
        } else {
            return set; 
        }
    }
    
    public Set<Word> getWordsWithLength(int length) {
        Set<Word> set = wordLength.get(length);
        if (set == null) {
            return new HashSet<>();
        } else {
            return set;
        }
    }
    
}
